File: smart-home/src/main/kotlin/io/github/tassiLuca/smarthome/coroutines/core/SensorSource.kt
package io.github.tassiLuca.smarthome.coroutines.core

import kotlinx.coroutines.flow.SharedFlow

/** A consumer of sensor events. */
interface SensorSourceConsumer<in E : SensorEvent> {
    /** Reacts to a sensor event. */
    suspend fun react(e: E)
}

/** A source of sensor events. */
interface SensorSource<out E : SensorEvent> {
    /** The flow of sensor events. */
    val sensorEvents: SharedFlow<E>
}

/** A detection performed by a sensing unit. */
sealed interface SensorEvent {
    /** The identifier of the sensing unit that generated the event. */
    val sourceUnit: String
}

/** A temperature detection. */
data class TemperatureEntry(
    override val sourceUnit: String,
    /** The temperature value in Celsius degrees. */
    val temperature: Double,
) : SensorEvent

/** A luminosity detection. */
data class LuminosityEntry(
    override val sourceUnit: String,
    /** The luminosity value in lux. */
    val luminosity: Double,
) : SensorEvent

File: smart-hub/src/main/scala/io/github/tassiLuca/smarthome/core/SensorSource.scala
package io.github.tassiLuca.smarthome.core

import io.github.tassiLuca.rears.Publisher

type Temperature = Double

/** A generic source of [[SensorEvent]]. */
trait SensorSource extends Publisher[SensorEvent]

/** A detection performed by a sensing unit. */
sealed trait SensorEvent(val name: String)
case class TemperatureEntry(sensorName: String, temperature: Temperature) extends SensorEvent(sensorName)
case class LuminosityEntry(sensorName: String, luminosity: Temperature) extends SensorEvent(sensorName)
